### PR TITLE
Custom annotations labels

### LINKED ISSUE #613: Allow user defined annotations for kafka channel dispatchers and receivers
**Problem**
I have a need to annotate dispatcher and receiver pods so that I can scrape logs and metrics with Datadog. While I could accomplish this with Fluentbit or Jaeger, all other resources within my cluster allow custom annotations. 

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)**
This feature is targeted at a Cluster Admin and/or developer.

**Exit Criteria**
As a cluster admin, I am able to add custom annotations to my dispatcher/receiver pods, validate that the pods have the annotations I expect and the annotations provide the behavior I expect.

### PR DESCRIPTION
**Pull request by travis-minke-sap**
Fixes #613

Exposes the ability to customize the Labels and Annotations of the Kubernetes resources (Services, Deployments, Pods) dynamically generated by the KafkaChannel controllers.

## Proposed Changes

- 🎁  Add new label/annotation fields to ConfigMap
- 🎁  Configure new label/annotation fields on "distributed" KafkaChannel resources
- 🎁  Fix compare to update new labels/annotations
- 🎁  Configure new label/annotation fields on "consolidated" KafkaChannel  (Dispatcher only - no Receiver!)
- 🧽  Refactor consolidated dispatcher Service creation to match new Deployment builder pattern.
- 🎁  Config README Documentation

**Release Note**

```release-note
KafkaChannel ConfigMap now supports custom Labels and Annotations for dynamically generated resources.
```

**Note:** The public facing docs already refer to the distributed configmap README which has been updated.

Holding for @devguyio (and/or other Consolidated KafkaChannel experts) to have a chance to review ; )
/hold

### CODE REVIEW

**Review by eric-sap — COMMENTED** [review]
Couple of questions; probably nothing to change

File: pkg/channel/consolidated/reconciler/controller/kafkachannel.go [hunk 1]
```diff
@@ -404,32 +414,50 @@ func (r *Reconciler) reconcileRoleBinding(ctx context.Context, name string, ns s
...
 				kc.Status.MarkServiceFailed("DispatcherServiceFailed", "Failed to create the dispatcher service: %v", err)
-				return svc, err
+				return err
 			}
-
-			return svc, err
+			return err
 		}
-
+		logger.Errorw("can't get dispatcher service", zap.Error(err), zap.String("namespace", dispatcherNamespace), zap.String("dispatcher-name", dispatcherName))
```
[the remark this answers is not in the document]
Comment (travis-minke-sap) [reply pkg/channel/consolidated/reconciler/controller/kafkachannel.go:443]: I would tend to agree, but in this case I was simply aligning with the error messaging that was already in the `reconcileDispatcher()` implementation above.  I chose consistency over "correctness" in this case.

File: pkg/channel/consolidated/reconciler/controller/kafkachannel.go [hunk 2]
```diff
@@ -346,10 +356,10 @@ func (r *Reconciler) reconcileDispatcher(ctx context.Context, scope string, disp
 		} else {
 			args.Replicas = *d.Spec.Replicas
 		}
-		want = resources.NewDispatcherBuilderFromDeployment(d.DeepCopy()).WithArgs(&args).Build()
+		want = resources.NewDispatcherDeploymentBuilderFromDeployment(d.DeepCopy()).WithArgs(&args).Build()
 
-		if !equality.Semantic.DeepEqual(want.Spec, d.Spec) {
-			logger.Infof("Dispatcher deployment changed; reconciling:\n%s", cmp.Diff(want.Spec, d.Spec))
+		if !equality.Semantic.DeepEqual(want.ObjectMeta, d.ObjectMeta) || !equality.Semantic.DeepEqual(want.Spec, d.Spec) {
```
[the remark this answers is not in the document]
Comment (travis-minke-sap) [reply pkg/channel/consolidated/reconciler/controller/kafkachannel.go:361]: Yeah, good eye - worth the conversation!   The reason it's ok to do the full ObjectMeta comparison is that the `want` value is based on the current live `d` version so they should be the same with the exception of the 4 ObjectMeta fields updated in the `Build()` operation.  I'll leave the conversation open for others to comment on, but I think it's ok like this.

**Comment by matzew** [comment]
If being deployed via an operator (that owns the manifests), using this feature might be potential conflict, if that changes the resources, the operator may roll them back, ?

File: pkg/channel/consolidated/reconciler/controller/kafkachannel.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (devguyio) [reply pkg/channel/consolidated/reconciler/controller/kafkachannel.go:361]: I _think_  it should be fine indeed due to what Travis mentioned.

**Comment by devguyio** [comment]
@matzew for the consolidated channel as you know, the dispatcher is owned by the controller, and with this PR only the dispatcher annotations/labels will be updated. are there any other resources the operator will be owning? I don't think the operator now owns the dispatcher.

@travis-minke-sap this lgtm, I'll just test it quickly today and I think it should be fine. thanks for doing this!!

**Comment by travis-minke-sap** [comment]
> @matzew for the consolidated channel as you know, the dispatcher is owned by the controller, and with this PR only the dispatcher annotations/labels will be updated. are there any other resources the operator will be owning? I don't think the operator now owns the dispatcher.

I am not too familiar with the Knative Operator, but my understanding follows what @devguyio is saying - These labels/annotations are only being placed on resources created/owned by the Controllers.  I would be surprised if the Operator was aware of, and modifying, those resources.
 
> @travis-minke-sap this lgtm, I'll just test it quickly today and I think it should be fine. thanks for doing this!!

Thanks - I did some super-basic manual testing of the consolidated channel channel as well (just fyi) but it would be good to have you confirm with your own checks!

**Comment by matzew** [comment]
> These labels/annotations are only being placed on resources created/owned by the Controllers.

More a generic warning. We had run into some issues w/ the improper ownership in the past, that's really all

**Comment by devguyio** [comment]
Works well thanks!
/lgtm

**Pull request merged**